File: backend/src/modules/plans/dto/update-plan.dto.ts
import { IsString, IsOptional, IsEnum, IsArray, IsBoolean, IsInt, IsNumber, Min, Max, MaxLength } from 'class-validator';
import { ApiProperty } from '@nestjs/swagger';
import { BillingInterval, BillingTiming } from './create-plan.dto';

export class UpdatePlanDto {
  @ApiProperty({ required: false })
  @IsString()
  @IsOptional()
  @MaxLength(100)
  name?: string;

  @ApiProperty({ required: false })
  @IsString()
  @IsOptional()
  @MaxLength(500)
  description?: string;

  @ApiProperty({ required: false, enum: BillingInterval })
  @IsEnum(BillingInterval)
  @IsOptional()
  billingInterval?: BillingInterval;

  @ApiProperty({ required: false, enum: BillingTiming, description: 'When to charge: IN_ADVANCE or IN_ARREARS' })
  @IsEnum(BillingTiming)
  @IsOptional()
  billingTiming?: BillingTiming;

  @ApiProperty({ required: false })
  @IsArray()
  @IsString({ each: true })
  @IsOptional()
  features?: string[];

  @ApiProperty({ required: false })
  @IsBoolean()
  @IsOptional()
  isActive?: boolean;

  @ApiProperty({ required: false, example: 30, description: 'Net payment terms in days' })
  @IsInt()
  @Min(0)
  @Max(365)
  @IsOptional()
  netPaymentTerms?: number;

  @ApiProperty({ required: false, example: 3, description: 'Grace period in days before draft invoices are finalized' })
  @IsInt()
  @Min(0)
  @Max(90)
  @IsOptional()
  invoiceGracePeriodDays?: number;

  @ApiProperty({ required: false, example: 1000, description: 'Usage cost threshold for progressive billing' })
  @IsNumber()
  @Min(0)
  @IsOptional()
  progressiveBillingThreshold?: number;
}

File: sdks/rust/src/api/types/update_plan_dto.rs
pub use crate::prelude::*;

/// Request type for API operation
#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct UpdatePlanDto {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(rename = "billingInterval")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub billing_interval: Option<UpdatePlanDtoBillingInterval>,
    /// When to charge: IN_ADVANCE or IN_ARREARS
    #[serde(rename = "billingTiming")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub billing_timing: Option<UpdatePlanDtoBillingTiming>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub features: Option<Vec<String>>,
    #[serde(rename = "isActive")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_active: Option<bool>,
    /// Net payment terms in days
    #[serde(rename = "netPaymentTerms")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub net_payment_terms: Option<f64>,
    /// Grace period in days before draft invoices are finalized
    #[serde(rename = "invoiceGracePeriodDays")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub invoice_grace_period_days: Option<f64>,
    /// Usage cost threshold for progressive billing
    #[serde(rename = "progressiveBillingThreshold")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub progressive_billing_threshold: Option<f64>,
}
